labels.Labels: reduce memory by storing as a single string

bboreham opened the pull request:
This commit adds an alternate implementation for `labels.Labels`, behind a build tag `stringlabels`.

Doc with background info: [link](https://docs.google.com/document/d/1fXrPI2kr-jUvkF9JkkA7aPk7UQYTG9Yvk6A-fSinqEU/edit?usp=sharing).

This builds on #11717, which created abstractions to let code access `labels.Labels` without knowing the internals.

The new data structure is held as a single `string`, containing all label/values like this:
    
        [len][name0][len][value0][len][name1][len][value1]...
    
The lengths are varint encoded so usually a single byte.
    
The previous structure using `[]Label` had 24 bytes of overhead for the slice and 16 for each label name and value; this one has 16 bytes overhead plus 1 for each name and value.

(I previously published this idea on [prometheus-developers](https://groups.google.com/g/prometheus-developers/c/Cz4PpzvVUxA/m/Xi3DxnYsHAAJ). ).  

In `Labels.Overwrite()` and `Labels.Hash()` we use an unsafe conversion from string to byte slice. `Overwrite` is explicitly unsafe, while in `Hash` this is purely to change the type so we can use the faster implementation.

As expected, benchmarks show some speedups and some slowdowns.  We are trading more CPU in a few cases for less in most cases and smaller overall heap size.  I think I can squeeze a bit more out of this design though.

Some notable improvements:
```
Labels_Hash/typical_labels_under_1KB-8                                                                          
128ns ± 0%         61ns ± 0%   -52.24%  (p=1.000 n=1+1)
RangeQuery/expr=rate(a_ten[1m])_+_rate(b_ten[1m]),steps=1000-8                                                 
4.56ms ± 0%       3.23ms ± 0%   -29.15%  (p=1.000 n=1+1)
LoadWAL/batches=100,seriesPerBatch=1000,samplesPerSeries=480,exemplarsPerSeries=24,mmappedChunkT=3800-8         
481MB ± 0%        388MB ± 0%   -19.51%  (p=1.000 n=1+1)
```

Full set (only run once before/after since they take 40 minutes): https://gist.github.com/bboreham/53626dde3a5f7cf732a23c1a745435fd

codesome commented:
![hooray-letsgo](https://user-images.githubusercontent.com/15064823/177575254-33f61249-2bac-404d-8fee-ab2217070aae.gif)

codesome commented:
/prombench main

prymitive commented:
I tried to test this patch to see what kind of memory savings it could give me but soon after startup I got this panic:

```
runtime error: slice bounds out of range [-11:]" stacktrace="goroutine 666554 [running]:
github.com/prometheus/prometheus/promql.(*evaluator).recover(0xc82cd2c120, 0xc66bf0e3b8, 0xc66bf0e3a8)
/src/promql/engine.go:950 +0x28f
panic({0x2d985e0, 0xccfa92f8c0})
/usr/local/go/src/runtime/panic.go:838 +0x207
github.com/prometheus/prometheus/model/labels.marshalLabelToSizedBuffer(0xc4b4a26420?, {0xcc69cd8500?, 0x39884f0?, 0x128?})
/src/model/labels/labels.go:720 +0x1b1
github.com/prometheus/prometheus/model/labels.marshalLabelsToSizedBuffer({0xc4b4a26400, 0x214b9?, 0x1f975?}, {0xcc69cd8500, 0x100, 0x128})
/src/model/labels/labels.go:711 +0x57
github.com/prometheus/prometheus/model/labels.(*SimpleBuilder).Overwrite(0xc00236f068, 0xc8250db528)
/src/model/labels/labels.go:847 +0xb3
github.com/prometheus/prometheus/tsdb/index.(*Decoder).Series(0xc00236f060, {0x7ec8a4b33841?, 0x37b1f840?, 0x10?}, 0xc000401c00?, 0xc8250db510)
/src/tsdb/index/index.go:1877 +0x2be
github.com/prometheus/prometheus/tsdb/index.(*Reader).Series(0xc00065ee40, 0x37b1f84, 0x0?, 0x2?)
/src/tsdb/index/index.go:1620 +0x127
github.com/prometheus/prometheus/tsdb.blockIndexReader.Series({{0x39a50b8?, 0xc00065ee40?}, 0xc002f38780?}, 0xc66bf0bff8?, 0x40d9a7?, 0x405000000000010?)
/src/tsdb/block.go:476 +0x55
github.com/prometheus/prometheus/tsdb.(*blockBaseSeriesSet).Next(0xc8250db4a0)
/src/tsdb/querier.go:445 +0x90
github.com/prometheus/prometheus/storage.(*genericMergeSeriesSet).Next(0xc82cd2c180)
/src/storage/merge.go:335 +0x288
github.com/prometheus/prometheus/storage.(*lazyGenericSeriesSet).Next(0x0?)
/src/storage/lazy.go:25 +0x68
github.com/prometheus/prometheus/storage.(*genericMergeSeriesSet).Next(0xc82cd2c1e0)
/src/storage/merge.go:335 +0x288
github.com/prometheus/prometheus/storage.(*lazyGenericSeriesSet).Next(0xcd5e0e40f0?)
/src/storage/lazy.go:25 +0x68
github.com/prometheus/prometheus/promql.expandSeriesSet({0x3997058, 0xc66aea9e30}, {0x3997bb8, 0xc79ce53260})
/src/promql/engine.go:893 +0x82
github.com/prometheus/prometheus/promql.checkAndExpandSeriesSet({0x3997058?, 0xc66aea9e30?}, {0x3997838?, 0xc756af7200})
/src/promql/engine.go:885 +0x78
github.com/prometheus/prometheus/promql.(*evaluator).eval(0xc82cd2c120, {0x3997838?, 0xc756af7200?})
/src/promql/engine.go:1542 +0x405
github.com/prometheus/prometheus/promql.(*evaluator).rangeEval(0xc82cd2c120, 0xc66bf0e300, 0xc66bf0e2d0, {0xc66bf0e190, 0x2, 0x0?})
/src/promql/engine.go:1034 +0x18f7
github.com/prometheus/prometheus/promql.(*evaluator).eval(0xc82cd2c120, {0x3997608?, 0xc82cd2c000?})
/src/promql/engine.go:1255 +0x4f6f
github.com/prometheus/prometheus/promql.(*evaluator).Eval(0x181b030d800?, {0x3997608?, 0xc82cd2c000?})
/src/promql/engine.go:965 +0xa7
github.com/prometheus/prometheus/promql.(*Engine).execEvalStmt(0xc001c06380, {0x3997058, 0xc66aea9a40}, 0xc64ef038f0, 0xc63cf429b0)
/src/promql/engine.go:696 +0xc97
github.com/prometheus/prometheus/promql.(*Engine).exec(0xc001c06380, {0x3997058, 0xc66aea9a40}, 0xc64ef038f0)
/src/promql/engine.go:595 +0x58a
github.com/prometheus/prometheus/promql.(*query).Exec(0xc64ef038f0, {0x3997058, 0xc66aea9920})
/src/promql/engine.go:197 +0x1f5
github.com/prometheus/prometheus/web/api/v1.(*API).queryRange(0xc002249cc0, 0xc9556ce200)
/src/web/api/v1/api.go:494 +0xe11
github.com/prometheus/prometheus/web/api/v1.(*API).Register.func2.1(0x398aa60?)
/src/web/api/v1/api.go:313 +0x14d
github.com/prometheus/prometheus/web/api/v1.(*API).Register.func1.1({0x398aa60, 0xc6636608c0}, 0x7f0780081590?)
/src/web/api/v1/api.go:288 +0x85
net/http.HandlerFunc.ServeHTTP(0x7f0780081590?, {0x398aa60?, 0xc6636608c0?}, 0xc63cf42960?)
/usr/local/go/src/net/http/server.go:2084 +0x2f
github.com/prometheus/prometheus/util/httputil.CompressionHandler.ServeHTTP({{0x3980820?, 0xc001738810?}}, {0x7f0780081590?, 0xc63cf42960?}, 0x397f600?)
/src/util/httputil/compression.go:91 +0x69
github.com/prometheus/prometheus/web.(*Handler).testReady.func1({0x7f0780081590?, 0xc63cf42960?}, 0x7f0780081500?)
/src/web/web.go:533 +0x39
net/http.HandlerFunc.ServeHTTP(0x7f0780081590?, {0x7f0780081590?, 0xc63cf42960?}, 0x66b36f?)
/usr/local/go/src/net/http/server.go:2084 +0x2f
github.com/prometheus/client_golang/prometheus/promhttp.InstrumentHandlerResponseSize.func1({0x7f0780081590?, 0xc63cf42910?}, 0xc9556ce200)
/home/builder/go/pkg/mod/github.com/prometheus/client_golang@v1.12.2/prometheus/promhttp/instrument_server.go:239 +0xc5
net/http.HandlerFunc.ServeHTTP(0x14e1090?, {0x7f0780081590?, 0xc63cf42910?}, 0xc66aea9710?)
/usr/local/go/src/net/http/server.go:2084 +0x2f
github.com/prometheus/client_golang/prometheus/promhttp.InstrumentHandlerDuration.func2({0x7f0780081590, 0xc63cf42910}, 0xc9556ce200)
/home/builder/go/pkg/mod/github.com/prometheus/client_golang@v1.12.2/prometheus/promhttp/instrument_server.go:84 +0xbf
net/http.HandlerFunc.ServeHTTP(0x3996ca0?, {0x7f0780081590?, 0xc63cf42910?}, 0x7f07a7951fe8?)
/usr/local/go/src/net/http/server.go:2084 +0x2f
github.com/prometheus/client_golang/prometheus/promhttp.InstrumentHandlerCounter.func1({0x3996ca0?, 0xcbc2eeff80?}, 0xc9556ce200)
/home/builder/go/pkg/mod/github.com/prometheus/client_golang@v1.12.2/prometheus/promhttp/instrument_server.go:117 +0xaa
github.com/prometheus/prometheus/web.setPathWithPrefix.func1.1({0x3996ca0, 0xcbc2eeff80}, 0xccd5c7ff00)
/src/web/web.go:792 +0x290
github.com/prometheus/common/route.(*Router).handle.func1({0x3996ca0, 0xcbc2eeff80}, 0xccd5c7fd00, {0x0, 0x0, 0xc8b7c672f5?})
/home/builder/go/pkg/mod/github.com/prometheus/common@v0.35.0/route/route.go:83 +0x2ae
github.com/julienschmidt/httprouter.(*Router).ServeHTTP(0xc000618180, {0x3996ca0, 0xcbc2eeff80}, 0xccd5c7fd00)
/home/builder/go/pkg/mod/github.com/julienschmidt/httprouter@v1.3.0/router.go:387 +0x82b
github.com/prometheus/common/route.(*Router).ServeHTTP(0x3996fb0?, {0x3996ca0?, 0xcbc2eeff80?}, 0x2031af?)
/home/builder/go/pkg/mod/github.com/prometheus/common@v0.35.0/route/route.go:126 +0x26
net/http.StripPrefix.func1({0x3996ca0, 0xcbc2eeff80}, 0xccd5c7fb00)
/usr/local/go/src/net/http/server.go:2127 +0x330
net/http.HandlerFunc.ServeHTTP(0xc6bc376c60?, {0x3996ca0?, 0xcbc2eeff80?}, 0xc5de5236c8?)
/usr/local/go/src/net/http/server.go:2084 +0x2f
net/http.(*ServeMux).ServeHTTP(0x3997058?, {0x3996ca0, 0xcbc2eeff80}, 0xccd5c7fb00)
/usr/local/go/src/net/http/server.go:2462 +0x149
go.opentelemetry.io/contrib/instrumentation/net/http/otelhttp.(*Handler).ServeHTTP(0xc00224e000, {0x3995560?, 0xcc55a8db20}, 0xccd5c7f500)
/home/builder/go/pkg/mod/go.opentelemetry.io/contrib/instrumentation/net/http/otelhttp@v0.32.0/handler.go:191 +0xd11
github.com/prometheus/prometheus/web.withStackTracer.func1({0x3995560?, 0xcc55a8db20?}, 0xc79ce52e80?)
/src/web/web.go:104 +0x97
net/http.HandlerFunc.ServeHTTP(0x0?, {0x3995560?, 0xcc55a8db20?}, 0x40d645?)
/usr/local/go/src/net/http/server.go:2084 +0x2f
net/http.serverHandler.ServeHTTP({0x3987870?}, {0x3995560, 0xcc55a8db20}, 0xccd5c7f500)
/usr/local/go/src/net/http/server.go:2916 +0x43b
net/http.(*conn).serve(0xc66aec0140, {0x3997058, 0xc000b965d0})
/usr/local/go/src/net/http/server.go:1966 +0x5d7
created by net/http.(*Server).Serve
/usr/local/go/src/net/http/server.go:3071 +0x4db
```

prymitive commented:
> @prymitive thanks for that; somehow the buffer it's trying to use is too small. Anything you can tell me about the query being executed when it crashed?

Sadly no, all I got was panics in the logs with multiple different indexes in the `slice bounds out of range` error, no idea which queries caused it.

roidelapluie commented:
The query executed is printed on the next startup when Prometheus reads the
active query logging

Le mer. 6 juil. 2022, 18:46, Łukasz Mierzwa ***@***.***> a
écrit :

> @prymitive <https://github.com/prymitive> thanks for that; somehow the
> buffer it's trying to use is too small. Anything you can tell me about the
> query being executed when it crashed?
>
> Sadly no, all I got was panics in the logs with multiple different indexes
> in the slice bounds out of range error, no idea which queries caused it.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/prometheus/prometheus/pull/10991#issuecomment-1176446876>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AACHHJUJDFKEDEDPTNEGL7LVSWZ65ANCNFSM52ZXDLTQ>
> .
> You are receiving this because you are subscribed to this thread.Message
> ID: ***@***.***>
>

prymitive commented:
> The query executed is printed on the next startup when Prometheus reads the active query logging Le mer. 6 juil. 2022, 18:46, Łukasz Mierzwa ***@***.***> a écrit :
> […](#)
> @prymitive <https://github.com/prymitive> thanks for that; somehow the buffer it's trying to use is too small. Anything you can tell me about the query being executed when it crashed? Sadly no, all I got was panics in the logs with multiple different indexes in the slice bounds out of range error, no idea which queries caused it. — Reply to this email directly, view it on GitHub <[#10991 (comment)](https://github.com/prometheus/prometheus/pull/10991#issuecomment-1176446876)>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AACHHJUJDFKEDEDPTNEGL7LVSWZ65ANCNFSM52ZXDLTQ> . You are receiving this because you are subscribed to this thread.Message ID: ***@***.***>

Prometheus never crashed, panic was recovered.
I will try to debug it more tomorrow.

bboreham commented:
> Prometheus never crashed, panic was recovered.

That seems like a good opportunity for improvement.  If a panic reaches the http handler then log the query.

Separately, can you say if this was the first panic?  If not can you post the stack trace for that one?

codesome commented:
Prombench results: There seems to be roughly 10% savings in RSS. Querying is similar, but a little more expensive in the prepare_time (which is where we get the samples from TSDB). The inner_eval, where most of the operations happen, is pretty much same. The different doesn't look concerning at the moment.

/prombench cancel

codesome commented:
/prombench cancel

prymitive commented:
I've added patch from #10995 and I see a few panics in logs.
They all seem to have range queries.
Here's an example:

```
component="query engine"
msg="runtime panic in parser"
expr="round(sum by(colo_name, interface) (increase(traffic_manager_local_interface_reactor_action_successful[30d]))) > 3"
err="runtime error: slice bounds out of range [-27:]"
stacktrace="goroutine 18108 [running]:
github.com/prometheus/prometheus/promql.(*evaluator).recover(0xc74ac63f80, {0x3997620, 0xc67660fac0}, 0xc3a33f0800, 0xc3a33f07f0)
	/src/promql/engine.go:950 +0x2a6
panic({0x2d985e0, 0xc2cc8bb818})
	/usr/local/go/src/runtime/panic.go:838 +0x207
github.com/prometheus/prometheus/model/labels.marshalLabelToSizedBuffer(0xc3c6cd5e20?, {0xc7007cdd10?, 0x44faf2?, 0xe3?})
	/src/model/labels/labels.go:720 +0x1b1
github.com/prometheus/prometheus/model/labels.marshalLabelsToSizedBuffer({0xc3c6cd5e00, 0x20ec2?, 0xc300020c58?}, {0xc7007cdd10, 0xe3, 0xe3})
	/src/model/labels/labels.go:711 +0x57
github.com/prometheus/prometheus/model/labels.(*SimpleBuilder).Overwrite(0xc007480a88, 0xc3a551b5d8)
	/src/model/labels/labels.go:847 +0xb3
github.com/prometheus/prometheus/tsdb/index.(*Decoder).Series(0xc007480a80, {0x7ed35b445cd1?, 0x6da28cd0?, 0x7e419d46efff?}, 0x7e419d374f00?, 0xc3a551b5c0)
	/src/tsdb/index/index.go:1877 +0x2be
github.com/prometheus/prometheus/tsdb/index.(*Reader).Series(0xc006d8acc0, 0x6da28cd, 0x7e41b0180c70?, 0x2?)
	/src/tsdb/index/index.go:1620 +0x127
github.com/prometheus/prometheus/tsdb.blockIndexReader.Series({{0x39a5098?, 0xc006d8acc0?}, 0xc004423680?}, 0x2343d6f?, 0x40d9a7?, 0x6040000002031b8?)
	/src/tsdb/block.go:476 +0x55
github.com/prometheus/prometheus/tsdb.(*blockBaseSeriesSet).Next(0xc3a551b550)
	/src/tsdb/querier.go:445 +0x90
github.com/prometheus/prometheus/storage.(*genericMergeSeriesSet).Next(0xc760f32120)
	/src/storage/merge.go:335 +0x288
github.com/prometheus/prometheus/storage.(*lazyGenericSeriesSet).Next(0x0?)
	/src/storage/lazy.go:25 +0x68
github.com/prometheus/prometheus/storage.(*genericMergeSeriesSet).Next(0xc760f32300)
	/src/storage/merge.go:335 +0x288
github.com/prometheus/prometheus/storage.(*lazyGenericSeriesSet).Next(0xc3a33ec1c8?)
	/src/storage/lazy.go:25 +0x68
github.com/prometheus/prometheus/promql.expandSeriesSet({0x3997038, 0xc183a9dda0}, {0x3997b98, 0xc1d75d4260})
	/src/promql/engine.go:893 +0x82
github.com/prometheus/prometheus/promql.checkAndExpandSeriesSet({0x3997038?, 0xc183a9dda0?}, {0x3997818?, 0xc6bd6c8a00})
	/src/promql/engine.go:885 +0x78
github.com/prometheus/prometheus/promql.checkAndExpandSeriesSet({0x3997038?, 0xc183a9dda0?}, {0x3997690?, 0xc156da4080?})
	/src/promql/engine.go:880 +0xf6
github.com/prometheus/prometheus/promql.(*evaluator).eval(0xc74ac63f80, {0x3997658?, 0xc3b574edb0?})
	/src/promql/engine.go:1344 +0x247b
github.com/prometheus/prometheus/promql.(*evaluator).rangeEval(0xc74ac63f80, 0xc3a33ee198, 0xc3a33ee168, {0xc3a33ee028, 0x2, 0x0?})
	/src/promql/engine.go:1034 +0x18f7
github.com/prometheus/prometheus/promql.(*evaluator).eval(0xc74ac63f80, {0x39975e8?, 0xc74ac63aa0?})
	/src/promql/engine.go:1255 +0x4f6f
github.com/prometheus/prometheus/promql.(*evaluator).rangeEval(0xc74ac63f80, 0x0, 0xc3a33ef2e0, {0xc392a49d40, 0x1, 0x0?})
	/src/promql/engine.go:1034 +0x18f7
github.com/prometheus/prometheus/promql.(*evaluator).eval(0xc74ac63f80, {0x3997658?, 0xc3b574ede0?})
	/src/promql/engine.go:1319 +0x2289
github.com/prometheus/prometheus/promql.(*evaluator).rangeEval(0xc74ac63f80, 0x0, 0xc3a33f0430, {0xc3a33f04f8, 0x2, 0x0?})
	/src/promql/engine.go:1034 +0x18f7
github.com/prometheus/prometheus/promql.(*evaluator).eval(0xc74ac63f80, {0x3997620?, 0xc67660fac0?})
	/src/promql/engine.go:1523 +0x14f2
github.com/prometheus/prometheus/promql.(*evaluator).Eval(0x181d85aa4d0?, {0x3997620?, 0xc67660fac0?})
	/src/promql/engine.go:965 +0xc5
github.com/prometheus/prometheus/promql.(*Engine).execEvalStmt(0xc001cb6080, {0x3997038, 0xc3b574ef30}, 0xc151e1a310, 0xc156d9e1e0)
	/src/promql/engine.go:645 +0x4fe
github.com/prometheus/prometheus/promql.(*Engine).exec(0xc001cb6080, {0x3997038, 0xc3b574ef30}, 0xc151e1a310)
	/src/promql/engine.go:595 +0x58a
github.com/prometheus/prometheus/promql.(*query).Exec(0xc151e1a310, {0x3997038, 0xc3b574ed50})
	/src/promql/engine.go:197 +0x1f5
github.com/prometheus/prometheus/rules.EngineQueryFunc.func1({0x3997038, 0xc3b574ed50}, {0xc6f5e21980?, 0xc3a33f0e18?}, {0x2384738?, 0x0?, 0x0?})
	/src/rules/manager.go:193 +0x7c
github.com/prometheus/prometheus/rules.(*AlertingRule).Eval(0xc3ad39d1e0, {0x3997038, 0xc3b574ed50}, {0x396de62?, 0x181d8517e14?, 0x0?}, 0xc0021e23e0, 0xc000e3a120, 0x0)
	/src/rules/alerting.go:328 +0xfe
github.com/prometheus/prometheus/rules.(*Group).Eval.func1({0x3997038, 0xc1e29e1c20}, 0xc3aea702d0, {0xc0b024f600?, 0xc0b024f600?, 0x0?}, 0xc3a33f1c68, 0x6, {0x39a7ce0, 0xc3ad39d1e0})
	/src/rules/manager.go:625 +0x457
github.com/prometheus/prometheus/rules.(*Group).Eval(0xc3aea702d0, {0x3997038, 0xc1e29e1c20}, {0x396de62?, 0xc000101678?, 0x0?})
	/src/rules/manager.go:710 +0x20d
github.com/prometheus/prometheus/rules.(*Group).run.func1()
	/src/rules/manager.go:363 +0x11b
github.com/prometheus/prometheus/rules.(*Group).run(0xc3aea702d0, {0x3996fc8, 0xc000050070})
	/src/rules/manager.go:400 +0x455
github.com/prometheus/prometheus/rules.(*Manager).Update.func1(0xc3c2d59b00?)
	/src/rules/manager.go:1010 +0xaf
created by github.com/prometheus/prometheus/rules.(*Manager).Update
	/src/rules/manager.go:1000 +0x3ad
"
```

prymitive commented:
Running a range query for `traffic_manager_local_interface_reactor_action_successful` over past 30d returns a list of time series and they're not all present over that time, that's about it really.

bboreham commented:
> I see a few panics in logs.

New theory: the problem is concurrent use of [this `SimpleBuilder`](https://github.com/prometheus/prometheus/pull/10991/files#diff-fb0f63e7d2d24fa7aa1c18b2619b10a07c821b38f45c2351458b00edef63f995R1781). The size of the buffer is calculated with one set of labels, then the labels change on another goroutine, and now the buffer is not big enough.

[side-note: I don't think there is a unit test of concurrent queries]

The `Decoder` is embedded in an [`index.Reader`](https://github.com/prometheus/prometheus/blob/96e20a3148cdd360770aa37182a19147a020f67f/tsdb/index/index.go#L1064-L1084) which I am reasonably sure is shared across multiple concurrent queries.

First thought is to move the stateful `SimpleBuilder` into the iterator-like structure which in the case of your panic is `blockBaseSeriesSet`.  However this implies changing the signature of the `Series` method to pass the `SimpleBuilder`.

Taking a more lateral approach, I can change `Labels` from a `string` to a `[]byte`, then use that as the "building" state as `Series` did before with the `[]Labels` slice. This is disappointing because it makes all `Labels` 8 bytes bigger for just this one case. 

Maybe I'll try both.

bboreham commented:
I managed to recreate the race I theorised in previous message in a unit test, and moved the state into `blockBaseSeriesSet` which makes the report go away.

Unfortunately the new test seems to run out of memory at CircleCI.  I tried reducing the concurrency from 10 to 4 but it still gets "signal: killed".  Maybe I can experiment with smaller data sets.

@prymitive if you can re-test that'd be great.

prymitive commented:
I did deploy it to the same test server and, so far, I don't see any panics, so it does look like the issue is resolved.

I've also captured heap pprof on 2.36.0 (+ #10782 & #10709 patches)  vs main+this patch.

v2.36.0
```
Showing nodes accounting for 36.67GB, 70.86% of 51.75GB total
Dropped 643 nodes (cum <= 0.26GB)
Showing top 10 nodes out of 90
      flat  flat%   sum%        cum   cum%
    6.80GB 13.14% 13.14%     6.80GB 13.14%  github.com/prometheus/prometheus/model/labels.(*Builder).Labels
    6.22GB 12.02% 25.16%    10.12GB 19.56%  github.com/prometheus/prometheus/tsdb/record.(*Decoder).Series
    5.27GB 10.19% 35.35%     5.27GB 10.19%  github.com/prometheus/prometheus/tsdb/chunkenc.NewXORChunk
    4.53GB  8.76% 44.11%     6.06GB 11.72%  github.com/prometheus/prometheus/tsdb.newMemSeries
    3.96GB  7.66% 51.76%     3.96GB  7.66%  github.com/prometheus/prometheus/tsdb/encoding.(*Decbuf).UvarintStr (inline)
    2.79GB  5.39% 57.15%     2.79GB  5.39%  github.com/prometheus/prometheus/model/textparse.(*PromParser).Metric
    1.94GB  3.74% 60.89%     1.94GB  3.74%  github.com/prometheus/prometheus/model/labels.NewBuilder
    1.74GB  3.36% 64.25%     1.74GB  3.36%  github.com/prometheus/prometheus/scrape.(*scrapeCache).addRef
    1.72GB  3.32% 67.57%     1.72GB  3.32%  github.com/prometheus/prometheus/tsdb/index.(*MemPostings).Delete
    1.70GB  3.29% 70.86%     1.70GB  3.29%  github.com/klauspost/compress/flate.(*dictDecoder).init
```

this patch:
```
Showing nodes accounting for 26.65GB, 65.24% of 40.85GB total
Dropped 658 nodes (cum <= 0.20GB)
Showing top 10 nodes out of 89
      flat  flat%   sum%        cum   cum%
    4.30GB 10.52% 10.52%     5.82GB 14.24%  github.com/prometheus/prometheus/tsdb.newMemSeries
    3.77GB  9.23% 19.75%     3.77GB  9.23%  github.com/prometheus/prometheus/model/labels.(*SimpleBuilder).Labels
    3.23GB  7.91% 27.65%     3.23GB  7.91%  github.com/prometheus/prometheus/model/labels.appendLabelTo
    3.07GB  7.52% 35.17%     3.07GB  7.52%  github.com/prometheus/prometheus/model/textparse.(*PromParser).Metric
    2.94GB  7.19% 42.36%     2.94GB  7.19%  github.com/prometheus/prometheus/tsdb/chunkenc.NewXORChunk
    2.33GB  5.70% 48.06%     5.57GB 13.63%  github.com/prometheus/prometheus/model/labels.(*Builder).Labels
    1.86GB  4.55% 52.61%     1.86GB  4.55%  github.com/prometheus/prometheus/scrape.(*scrapeCache).addRef
    1.81GB  4.44% 57.05%     1.81GB  4.44%  compress/flate.(*dictDecoder).init
    1.71GB  4.19% 61.24%     1.71GB  4.19%  github.com/prometheus/prometheus/tsdb/index.(*MemPostings).Delete
    1.63GB  4.00% 65.24%     1.63GB  4.00%  github.com/prometheus/prometheus/scrape.(*scrapeCache).trackStaleness
```

prymitive commented:
Somehow `external_labels` are missing from `/api/v1/status/config` response, as if marshaling to JSON didn't work properly.

prymitive commented:
This is go alloc size (`avg_over_time(go_memstats_alloc_bytes[30m])`) on a test instance. We have:
- v2.36.0 + patches (#10782 & #10709 patches) running initially
- at 08:50 main+this patch starts
- at 13:20 it's reverted back to v2.36.0

![image](https://user-images.githubusercontent.com/114288/178314036-3be898bc-2a90-403f-9d2b-5ea3c4dc2ca4.png)

Had to revert because problem with external_labels mentioned above breaks thanos - it tries to read external_labels from config via the API but gets nothing so it exits.

bboreham commented:
> Somehow `external_labels` are missing from `/api/v1/status/config` response, as if marshaling to JSON didn't work properly.

I fixed a flaw with marshalling to YAML, which I think you meant.
Also added a test for JSON.

prymitive commented:
Thanks! I can confirm that it fixed external_labels issue

prymitive commented:
I've deployed this patch to two instances today, that's on top of v2.37.0-rc1 (+ #10782 & #10709 patches since we have those on top of current v2.36.0). It's working fine after a few hours so I'll keep those two instance running this patch.
Comparing go allocation bytes between these two shows a decent drop.
- 94 -> 76GB (peak) for instance with 26.8M time series
- 56 -> 46GB (peak) for instance with 16.4M time series
That's about 18-20% less, which is amazing.

![image](https://user-images.githubusercontent.com/114288/178745095-6c918459-7013-4624-8da8-d39cd2721bc1.png)
![image](https://user-images.githubusercontent.com/114288/178745151-3b9c4918-9227-4c37-bc35-20fe711cdc68.png)

prymitive commented:
FYI both instances are working perfectly fine and memory usage reduction is looking great.
26.8M time series instance:
![image](https://user-images.githubusercontent.com/114288/178978588-130c88da-0eec-4105-875f-8fe306fadaa0.png)

kwiesmueller commented on model/labels/labels.go:
```diff
@@ -281,68 +290,83 @@ func (ls Labels) Has(name string) bool {
-	m := make(map[string]string, len(ls))
-	for _, l := range ls {
-		m[l.Name] = l.Value
+	m := make(map[string]string, len(ls.data)/10)
```
Where does that 10 come from?
From my understanding so far it should be 2+something for the length information.

bboreham replied:
This is to size the map, which will contain all label names and values.  It would be correct if all label names and values are length 4; the map size will be a bit of an over-estimate in most cases, which is ok.  Probably deserving a comment.

kwiesmueller commented on model/labels/labels.go:
```diff
@@ -359,39 +383,212 @@ func FromStrings(ss ...string) Labels {
+}
+
+// Len returns the number of labels; it is relatively slow.
+func (ls Labels) Len() int {
```
Depending on how much it is needed it might be worth to track the len in a separate field.

bwplotka replied:
We can iterate over it too in later PRs (:

kwiesmueller commented on scrape/scrape.go:
```diff
@@ -619,7 +620,7 @@ func verifyLabelLimits(lset labels.Labels, limits *labelLimits) error {
 	met := lset.Get(labels.MetricName)
 	if limits.labelLimit > 0 {
-		nbLabels := len(lset)
+		nbLabels := lset.Len()
```
Regarding the cost of len.
As this happens on every append it might really be worth making it cheaper.

kwiesmueller replied:
Also as an idea for future optimization.
How about we provide the limits to your new label processing code and validate labels while they are being written instead of iterating another time afterwards?

bboreham replied:
On my reading of the code it does not happen on every append. It is called from the branch if the series is not in cache, i.e. if it has never been seen before.

https://github.com/prometheus/prometheus/pull/10991/commits/5a472f285ae81523ce28c8b0f52ea80c69ea227e#diff-4535d6172e0a9ab4e0879d71ef67ab69023f6eda9b44100343a6afa04fc8925fR1559

bboreham commented on model/labels/labels.go:
```diff
@@ -116,68 +146,46 @@ func (ls *Labels) UnmarshalYAML(unmarshal func(interface{}) error) error {
+	if on {
+		b.Keep(names...)
+	} else {
+		b.Del(MetricName)
```
This matches line 128 in the old code: `(on || v.Name != MetricName)`

bboreham commented on model/labels/labels.go:
```diff
@@ -205,73 +216,71 @@ func (ls Labels) HashWithoutLabels(b []byte, names ...string) (uint64, []byte) {
+	for i := 0; i < len(ls.data); {
+		var lName string
+		lName, i = decodeString(ls.data, i)
+		_, i = decodeString(ls.data, i)
```
We don't need it in the case `lName == name`, true.

bboreham reviewed (commented):
Thank you for your comments, @kwiesmueller

prymitive commented:
Do you it would be feasible to further compress the data? Maybe with some dictionary based compression method (although that seem to require a pre trained dictionary, for example with zstd).
Maybe the easiest option would be to substitute a few well known labels with short magic values, at least job and instance labels?

roidelapluie commented:
And `__name__`

bboreham commented:
Thanks; I had not thought of dictionary-style compression.  I guess the decompression would happen in `Range()` etc.?

Couple of ideas I had considered, as small modifications to this current PR:
1. Pull `__name__` out to a separate string.  This saves 9 bytes for the label but adds 16 bytes for a string header. Maybe we can "intern" those strings, then we can use a pointer-to-string, but now you have a global lock on label creation.
2. Divide labels into "target" and "instance" sets - everything for a single scrape gets the same target labels. Now sharing that bit is easier, but again the header costs an extra 16 bytes. All Labels methods get more complicated.

I think the limit, in terms of complexity and space savings, is a trie data structure.

There are many possible trade-offs; really I needed to do all the groundwork in #10887 just to be able to experiment.

prymitive commented:
> Divide labels into "target" and "instance" sets - everything for a single scrape gets the same target labels.

If `honor_labels: true` is set then each target can return a completely unique set of labels. But that wouldn't break anything, just mean you'll sometimes have an empty "target" set.
I would probably worry more about `metric_relabel_configs` as people sometimes do crazy, or at least unnecessary, things there from my experience. Like setting `job_name` to foo just to add a `metric_relabel_configs` rule that sets job to `bar`.

prymitive commented:
>     1. Pull `__name__` out to a separate string.  This saves 9 bytes for the label but adds 16 bytes for a string header. Maybe we can "intern" those strings, then we can use a pointer-to-string, but now you have a global lock on label creation.

Labels have a limited set of chars to use:

`Label names may contain ASCII letters, numbers, as well as underscores. They must match the regex [a-zA-Z_][a-zA-Z0-9_]*. Label names beginning with __ are reserved for internal use.`

So I was curious if we could just store `__name__` as a single char from the range that's not allowed by the regexp above.
Instead of `"\b__name__\x06ALERTS"` have `"\x01!"` (if we use `!` for `__name__`).

roidelapluie commented:
That is a limitation of Prometheus, not of the TSDB, AFAIK.

prymitive commented:
> That is a limitation of Prometheus, not of the TSDB, AFAIK.

Good point.
Maybe in the future, once this is all merged, further refactor might establish some interfaces around the encoding code. So that Prometheus can pass code that takes advantage of that, without forcing it on the entire tsdb.

prymitive commented:
I was reading on string interning in Go a bit:
https://artem.krylysov.com/blog/2018/12/12/string-interning-in-go/
https://commaok.xyz/post/intern-strings/
https://github.com/golang/go/issues/5160

There are some existing libraries providing interning - like https://github.com/lkarlslund/stringdedup - but they all seem to come with a lot of warnings, so overall it seems like a great but complicated solution.

roidelapluie commented:
> I was reading on string interning in Go a bit: https://artem.krylysov.com/blog/2018/12/12/string-interning-in-go/ https://commaok.xyz/post/intern-strings/ [golang/go#5160](https://github.com/golang/go/issues/5160)
> 
> There are some existing libraries providing interning - like https://github.com/lkarlslund/stringdedup - but they all seem to come with a lot of warnings, so overall it seems like a great but complicated solution.

This requires Go 1.18 on x86 / x64. Please let me know your experiences.

That will be one key issue. When we moved to mmap in 2.19., we did lose support for OpenBSD. In this case, we could lose more archs, which I would like to avoid.

bboreham commented:
> I was reading on string interning in Go a bit:

I have too, but I cannot see how it could be relevant to this PR.  I do not expect two `Labels` to have the same string representation very often, except in cases where we are looking up the one in the TSDB head, which logically about the same as interning.

Are you mentioning it in regard to a rather different design, or did I miss something?

prymitive commented:
`Labels.String()`  will be unique, but individual `Label` name & values will be repeated, and that's where deduplicating opportunity lies IMHO.
`__name__` alone is repeated for every time series.

roidelapluie commented:
> `Labels.String()`  will be unique, but individual `Label` name & values will be repeated, and that's where deduplicating opportunity lies IMHO.
> `__name__` alone is repeated for every time series.

My understanding is that this pull request produces a single string for a label set.

prymitive commented:
Sorry, I though interning was mentioned here as an alternative, seems I got threads confused.
Yes, this keeps everything as a single string and so interning doesn't apply to this implementation.

prymitive commented on model/labels/labels.go:
```diff
@@ -205,73 +216,71 @@ func (ls Labels) HashWithoutLabels(b []byte, names ...string) (uint64, []byte) {
-	res := make(Labels, len(ls))
-	copy(res, ls)
-	return res
+	buf := append([]byte{}, ls.data...)
```
Since strings are immutable is this really needed?
I would think that `Labels{data: ls.data}` would be enough and there's no need to allocate anything.

bboreham replied:
Strings are immutable except when people use `unsafe`...

It is used in places like this (on current `main`): https://github.com/prometheus/prometheus/blob/3947238ce0e3ff7e7c6e9502f73275cdbc233054/tsdb/querier.go#L509-L510

which I replaced with a call to `Labels.Copy()`.

To expand a little, we are re-using the buffer inside `labels.Labels` to reduce memory allocations.
This is non-`unsafe` when you do it with slices, but you still have to be extremely careful to copy in all the right places.

Probably there is a cleaner way where we pass some higher-level iterator object around.

bboreham commented:
I finally tidied up my design doc; it's not hugely detailed but should give some idea besides reading the code.
https://docs.google.com/document/d/1fXrPI2kr-jUvkF9JkkA7aPk7UQYTG9Yvk6A-fSinqEU/

prymitive commented:
Looks like there's a fair number of non-trivial conflicts due to some changes made in v2.39.0 release.

bboreham commented:
I rebased against main and fixed everything up so tests pass.

Because some changes have been merged in other PRs, this PR is a bit smaller:
Previously: Files changed 76; Lines +1,874 −1,705 
Now: Files changed 64; Lines +1,299 −827

bwplotka reviewed (commented):
Amazing work. New implementation LGTM. Thanks! Comments about the interface are here: https://github.com/prometheus/prometheus/pull/10887#pullrequestreview-1182414843

bboreham commented:
/prombench main

codesome commented:
/prombench main

roidelapluie commented:
/prombench cancel

Le sam. 3 déc. 2022, 12:10, PrometheusBot ***@***.***> a
écrit :

> Benchmark tests are running for 5 days! If this is intended ignore this
> message otherwise you can cancel it by commenting: /prombench cancel
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/prometheus/prometheus/pull/10991#issuecomment-1336136555>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AACHHJTF4OBQMYQLNZCG5B3WLMTDPANCNFSM52ZXDLTQ>
> .
> You are receiving this because you commented.Message ID:
> ***@***.***>
>

prymitive commented:
> Pull __name__ out to a separate string. This saves 9 bytes for the label but adds 16 bytes for a string header. Maybe we can "intern" those strings, then we can use a pointer-to-string, but now you have a global lock on label creation.

As @bboreham notes any global interning of strings would require a centralized storage of strings and so lock, which is always a performance risk.

One way of getting around this could be to use postings index as the source of interned strings.
The MemPostings used by HEAD is a `label key`->`label value`->`series refs` mapping and it should contain all possible known label pairs.

I don't think one can get the actual key instance from a map without ranging over it, so would probably have to be refactored into something like:

```go
map[string]struct{
  key       string
  values map[string]struct{
      value string
      refs []storage.SeriesRef
  }
}
```

to be able to get interned key & value string from it.
Plus, due to all the overhead of storing all individual strings, even interned, interning might require more memory than one big string with all labels per time series.

npordash commented:
Hi @bboreham 

I am wondering if there has been any testing done internally at Grafana against various real workloads that could be shared? I read through the linked design document and if I'm understanding things correctly any workload where the average label name/value size is smaller than a string pointer is obviously going to come out a head compared to alternatives like string interning.

If I look at our workloads this is quite often not the case though. At least within the context of Kubernetes it is quite common for things like namespace, application, pod, node, etc to exceed 16 bytes. We're currently amortizing the cost of these through string interning that @vjsamuel mentioned in #11833, which would no longer be possible with this proposal, and I am a bit concerned if this might negatively impact overall memory usage in these cases.

roidelapluie commented:
I think it would be interesting to merge this pull request with build tags.
We could then have users testing both and get feedback from the field.

Le mar. 10 janv. 2023, 21:34, Nick Pordash ***@***.***> a
écrit :

> Hi @bboreham <https://github.com/bboreham>
>
> I am wondering if there has been any testing done internally at Grafana
> against various real workloads that could be shared? I read through the
> linked design document and if I'm understanding things correctly any
> workload where the average label name/value size is smaller than a string
> pointer is obviously going to come out a head compared to alternatives like
> string interning.
>
> If I look at our workloads this is quite often not the case though. At
> least within the context of Kubernetes it is quite common for things like
> namespace, application, pod, node, etc to exceed 16 bytes. We're currently
> amortizing the cost of these through string interning that @vjsamuel
> <https://github.com/vjsamuel> mentioned in #11833
> <https://github.com/prometheus/prometheus/pull/11833>, which would no
> longer be possible with this proposal, and I am a bit concerned if this
> might negatively impact overall memory usage in these cases.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/prometheus/prometheus/pull/10991#issuecomment-1377813889>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AACHHJUSG7ITGMKKW2INUM3WRXBURANCNFSM52ZXDLTQ>
> .
> You are receiving this because you commented.Message ID:
> ***@***.***>
>

bboreham commented:
I have extended the [design doc](https://docs.google.com/document/d/1fXrPI2kr-jUvkF9JkkA7aPk7UQYTG9Yvk6A-fSinqEU/edit#) with some ideas to reduce repeated values within the single string.

bboreham commented:
> within the context of Kubernetes it is quite common for things like namespace, application, pod, node, etc to exceed 16 bytes

First of all, I think we have to be careful not to assume that our environment represents the situation for all Prometheus users. Any particular design will favour some set of data and look less well with another.

Second, each label is a pair (name, value) and the name ("pod", "node", etc.,) is another string costing another 16 bytes. You need to be working with pod names over 27 bytes, namespace names over 22 bytes, etc., to come out ahead of the implementation in this PR. 

(It does happen -- for example our node names are around 52 bytes, but this is balanced by the enormous number of labels like "le", "job", "interface", etc., which are shorter.)

bboreham commented:
> I think it would be interesting to merge this pull request with build tags.

Sure; it would be helpful if CI tests ran also with the tag so we catch people using the slice-ness of `Labels`.

> We could then have users testing both and get feedback from the field.

Do you think Prometheus would publish two images with/without?

roidelapluie commented:
Great suggestion, I agree that adding tests with the tags would be beneficial to catch any issues that may arise with using the slice-ness of Labels.
One possible way to achieve this is by adding an extra go test command in the CI pipeline located here: https://github.com/prometheus/prometheus/blob/cb2be6e62ff80ba84b639e4251044214e185886f/.github/workflows/ci.yml#L18

As for publishing two images, I think it's worth considering but it does not seem like it's a precedent we want to set. For most use cases, we prefer experimental flags, which is not possible in this case. However, for power users, I would expect them to rebuild Prometheus from source anyway. What are your thoughts on this?

bboreham commented:
> However, for power users, I would expect them to rebuild Prometheus from source anyway. What are your thoughts on this?

Agreed; I was just exploring the extent of your suggestion.

roidelapluie commented:
Do you need help with the build tag and the CI?

bboreham commented:
I haven’t got to it; I would certainly appreciate the assistance.

bboreham commented:
With @roidelapluie assistance I have made the suggested change, so the new code is available with a build tag `stringlabels`, and CI runs tests with that tag to check nothing breaks.

I broke out #12014 but the same commit is included in this PR.

roidelapluie commented:
Thanks!

codesome commented:
Amazing! How do we plan to handle releases for this? Will the release binary have these new labels implementation used?

roidelapluie commented:
So, now we know that tests pass with this. I propose we keep CI running with this for this release cycle. In the meantime, it would be great to see if we can find users to use a prometheus release (or mimir or thanos) built with this version and this tag (it would be easier if we first release this).

In the release after the next release, if it is positive for the testers, we replace the original implementation with the new one.

bboreham commented:
I looked through the release instructions and Makefile, but I didn't see a way to set Go tags.

`promu` will apply tags listed in its config; would we create an alternate copy of that to do the alternate build?

bboreham commented:
Update: I built a Docker image under my own account, and have been running it in a dev environment for the past 3 days. 
This suggested some further optimisation - #12048

djluck commented:
Our observation with our own cluster (normal memory consumption ~50GB):
![image](https://user-images.githubusercontent.com/1540536/226457083-5f5972fe-6b1e-4559-a5f9-8f244d481861.png)

This graph shows the absolute amount of memory consumed by two nodes in our cluster, one running the patch version and one running v2.29.1 (lol we are behind on our upgrades). We can see that memory consumption can be as much as 10GB less (although some peaks are almost the same). The period at the start of the graph shows where the two nodes were running the same version (identical mem. consumption).

![image](https://user-images.githubusercontent.com/1540536/226457802-a2f00533-24d9-4c81-bfd2-7c97407efac9.png)
This graph shows the % of memory saved by running this patch version, ranging around 5-25%.

bboreham commented:
Thanks for posting your findings. 
There is now an image published from Prometheus CI: `prom/prometheus:v2.43.0-rc.1_stringlabels`

der-eismann commented:
Unfortunately for us there don't seem to be any memory savings compared to the instance running v2.42.0 before, however CPU consumption seems to have calmed down a bit.

![screenshot-20230323-105851](https://user-images.githubusercontent.com/10406217/227168731-8c2bdbb3-f081-4281-b7a2-8c6bcee42896.png)

bboreham commented:
Thanks for the data, @der-eismann.  

A saying of mine: "there is no single metric for memory which shows what is going on".
The fact that yours hits the red limit suggests you are experiencing some paging.  It might even have got worse.
Suggest plotting that, and/or the Go heap size.

`rate(container_memory_failures_total{..., failure_type="pgmajfault"}[5m]`
`go_memstats_heap_inuse_bytes` (`go_memstats_next_gc_bytes` is more stable, but requires more understanding of Go internals to read)

der-eismann commented:
Happy to help @bboreham! You're completely correct with your saying and thanks for suggesting some useful metrics - I'm pretty sure the memory allocation is not perfect, it's more of a "it's not dying, so let's keep it :man_shrugging:". 

The `container_memory_failures_total{..., failure_type="pgmajfault"}` seem much more stable, or at least not growing as much as before, though that is hard to see in the rate due to the spikes.
![screenshot-20230323-141925](https://user-images.githubusercontent.com/10406217/227217123-02fb3467-e616-4f5e-b93a-73b30f70dd26.png)
![screenshot-20230323-141909](https://user-images.githubusercontent.com/10406217/227217126-493090df-b205-4de5-b28f-76efe22db6c3.png)

`go_memstats_heap_inuse_bytes` seems to be a bit lower while `go_memstats_next_gc_bytes` stayed the same, unfortunately we only keep those metrics for 48h.
![screenshot-20230323-143023](https://user-images.githubusercontent.com/10406217/227219307-4a6e14fb-4df5-4acd-8b5f-f2b3d7b98df3.png)
![screenshot-20230323-143018](https://user-images.githubusercontent.com/10406217/227219311-dab3e068-e1b1-417f-b3b7-c51261e867dc.png)
![screenshot-20230323-142945](https://user-images.githubusercontent.com/10406217/227219313-bc3bb51c-6324-4bca-b402-922e448b0668.png)
![screenshot-20230323-142940](https://user-images.githubusercontent.com/10406217/227219316-9a0987ae-a94c-4920-b142-5adace6d652d.png)

Let me know if you need any other metrics!

hamishforbes commented:
In my environment I'm seeing modest (possibly within margin of error tbh) memory improvements, but also a CPU usage reduction!

![Screenshot 2023-03-23 at 20 00 41](https://user-images.githubusercontent.com/1282135/227359185-0762290b-7b94-404f-9dbf-ee80a2b85657.png)

![Screenshot 2023-03-23 at 20 03 44](https://user-images.githubusercontent.com/1282135/227359208-cf2cfe5f-d9ef-44b1-8c4b-2f032eacf22c.png)

Compared to `2.43.0`.
This instance has around 1m series, it's a kube-prometheus-stack deployment that remote writes to Mimir, so the usage pattern is probably a bit different from normal.
There's almost no querying, just recording rules and some 'local' querying (from autoscalers etc).
Mostly just scraping and forwarding.
Not sure if that makes a difference

bboreham commented:
@der-eismann I feel I need to explain your first chart: I suggested `rate(5m)`, but when zoomed out to 2 weeks this window was too small to show all the data. Your screengrab is 1000 pixels wide and 2 weeks / 1000 is ~20 minutes, so with a 5-minute window the other 15 minutes of each pixel is not included in the calculation. If you try `[20m]` or `[30m]` it will likely be easier to see the pattern.

@hamishforbes indeed the specific pattern of usage can make a big difference - I recently realised that _dropped_ targets hold on to huge `Labels` just in case you want to view them to diagnose why they were dropped. Thanks for sharing your data.

npordash commented:
We've got an internal system I've got to successfully build with `stringlabels` and in at least one use-case so far that I've tested we're seeing a ~20% reduction in cpu (read+write workloads) and a ~15% reduction in memory.

AFAICT the cpu reduction is largely coming from the reduced cost to construct labels on both the read and write side. The write side is translating from protobuf into a shared `labels.Labels` and `labels.ScatchBuilder` before working with the appender while the read side appears to be mostly from https://github.com/prometheus/prometheus/blob/v2.43.0/tsdb/querier.go#L531.

I did run into a few hiccups in trying to get our code working with both `stringlabels` and `!stringlabels` builds which I've filed a separate issue for: https://github.com/prometheus/prometheus/issues/12230.

Thanks for all of the hard work! So far this is looking very promising :)

Stono commented:
Hello
I'm just here to give you a round of applause 

<img width="1649" alt="Screenshot 2023-04-27 at 12 55 21" src="https://user-images.githubusercontent.com/1486729/234854634-f2901ad3-f76e-4788-b6db-b767b02caae8.png">

👏

djluck commented:
An update after rolling this out to our infra that stores 70 million time series. We saw the following resource use reduction after rolling this out:
- 190 -> 145 cpu cores
- 1.3TB -> 1TB memory

I just want to thank everyone involved in this change, it's made a remarkable improvement.
